src/aliceVision/sfm/pipeline/ReconstructionEngine.cpp

```diff
@@ -0,0 +1,77 @@
+// This file is part of the AliceVision project.
+// Copyright (c) 2019 AliceVision contributors.
+// This Source Code Form is subject to the terms of the Mozilla Public License,
+// v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "ReconstructionEngine.hpp"
+
+#include <aliceVision/feature/RegionsPerView.hpp>
+#include <aliceVision/sfm/pipeline/regionsIO.hpp>
+
+
+namespace aliceVision {
+namespace sfm {
+
+
+void retrieveMarkersId(sfmData::SfMData& sfmData)
+{
+    std::set<feature::EImageDescriberType> allMarkerDescTypes;
+#if ALICEVISION_IS_DEFINED(ALICEVISION_HAVE_CCTAG)
+    allMarkerDescTypes.insert(feature::EImageDescriberType::CCTAG3);
+    allMarkerDescTypes.insert(feature::EImageDescriberType::CCTAG4);
+#endif
+    if (allMarkerDescTypes.empty())
+        return;
+
+    std::set<feature::EImageDescriberType> usedDescTypes = sfmData.getLandmarkDescTypes();
+
+    std::vector<feature::EImageDescriberType> markerDescTypes;
+    std::set_intersection(allMarkerDescTypes.begin(), allMarkerDescTypes.end(),
+        usedDescTypes.begin(), usedDescTypes.end(),
+        std::back_inserter(markerDescTypes));
+
+    std::set<feature::EImageDescriberType> markerDescTypes_set(markerDescTypes.begin(), markerDescTypes.end());
+
+    if(markerDescTypes.empty())
+        return;
+
+    // load the corresponding view regions
+    feature::RegionsPerView regionPerView;
+    std::set<IndexT> filter;
+    // It could be optimized by loading only the minimal number of desc files,
+    // but as we are only retrieving them for markers, the performance impact is limited.
+    if (!sfm::loadRegionsPerView(regionPerView, sfmData, sfmData.getFeaturesFolders(), markerDescTypes, filter))
+    {
+        ALICEVISION_THROW_ERROR("Error while loading markers regions.");
+    }
+    for (auto& landmarkIt : sfmData.getLandmarks())
+    {
+        auto& landmark = landmarkIt.second;
+        if (landmark.observations.empty())
+            continue;
+        if (markerDescTypes_set.find(landmark.descType) == markerDescTypes_set.end())
+            continue;
+        landmark.rgb = image::BLACK;
+
+        const auto obs = landmark.observations.begin();
+        const feature::Regions& regions = regionPerView.getRegions(obs->first, landmark.descType);
+        const feature::CCTAG_Regions& cctagRegions = dynamic_cast<const feature::CCTAG_Regions&>(regions);
+        const auto& d = cctagRegions.Descriptors()[obs->second.id_feat];
+        for (int i = 0; i < d.size(); ++i)
+        {
+            if (d[i] == 255)
+            {
+                ALICEVISION_LOG_TRACE("Found marker: " << i << " (landmarkId: " << landmarkIt.first << ").");
+                landmark.rgb.r() = i;
```

Comment:
why don't we just add another field to landmark like `markerId` always initialized to a dummy value (for all the other features that are not markers)?
Using the color attribute is a kind of workaround. Also, it is limited to 255 values, which is limiting in case other types of markers will be integrated in the future (eg. aruco &C).
Ok we need to change all the IO functions of sfmdata but I think it's better than "hiding" data in the same structure. And I think we can still preserve backcompatibility with other sfmdata files that has not that field.

Reply:
As I explained in the PR summary, it will be a waste of space as most of the landmarks are not markers. I agree that it looks hackish. I was considering using an union between `RGBColor` and `int` but that represents many changes in IOs and creates complexity.
The correct long-term solution is probably to have landmarks per descType, but that's a major change that I cannot afford for now.